contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import {IForwarder} from "./interfaces/IForwarder.sol";
+import {ConfirmedOwner} from "../shared/access/ConfirmedOwner.sol";
+import {TypeAndVersionInterface} from "../interfaces/TypeAndVersionInterface.sol";
+import {AccessControllerInterface} from "../shared/interfaces/AccessControllerInterface.sol";
+
+contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterface {
+    error ReentrantCall();
+
+    struct HotVars {
+        bool reentrancyGuard; // guard against reentrancy
+    }
+
+    /// @notice The contract to control addresses that are allowed to verify reports
+    AccessControllerInterface public s_accessController;
+
+    HotVars internal s_hotVars; // Mixture of config and state, commonly accessed
+
+    mapping(bytes32 => address) internal s_reports;
+
+    constructor(AccessControllerInterface accessController) ConfirmedOwner(msg.sender) {
+        s_accessController = accessController;
+    }
+
+    function splitSignature(
+        bytes memory sig
+    ) public pure returns (bytes32 r, bytes32 s, uint8 v) {
+        require(sig.length == 65, "invalid signature length");
+
+        assembly {
+            /*
+            First 32 bytes stores the length of the signature
+
+            add(sig, 32) = pointer of sig + 32
+            effectively, skips first 32 bytes of signature
+
+            mload(p) loads next 32 bytes starting at the memory address p into memory
+            */
+
+            // first 32 bytes, after the length prefix
+            r := mload(add(sig, 32))
+            // second 32 bytes
+            s := mload(add(sig, 64))
+            // final byte (first byte of the next 32 bytes)
+            v := byte(0, mload(add(sig, 96)))
+        }
+
+        // implicitly return (r, s, v)
+    }
+
+    function splitReport(
+        bytes memory rawReport
+    ) public pure returns (bytes32 workflowId, bytes32 workflowExecutionId) {
+        require(rawReport.length > 64, "invalid report length");
+        assembly {
+            workflowId := mload(add(rawReport, 4))
+            workflowExecutionId := mload(add(rawReport, 36)) // 4 + 32
+        }
+    }
+
+    // send a report to targetAddress
+    // report(workflowRunID, targetAddress, report, signatures)
+    function report(address targetAddress, bytes calldata data, bytes[] calldata signatures) external nonReentrant returns (bool) {
+    	// TODO: maybe forward as foo(report, calldata...) so we can validate sigs separately
+
+        // data is an encoded call with the selector prefixed: (bytes4 selector, bytes report, ...)
+        // we are able to partially decode just the first param, since we don't know the rest
+        bytes memory rawReport = abi.decode(data[4:], (bytes));
```

Comment:
@bolekk the limitations/assumptions here are that `report = concat(workflow id, execution id, data...)` and that the target method call can be anything but the report needs to be passed as first parameter, e.g. `myFoo(report, meta, other data)`

Reply:
Yeah I think that works, I will change my encoder accordingly.
We are also assuming that len(workflowID) == len(executionID) == 32 bytes, right?

Reply:
yep!